near-contract-standards/src/event.rs

```diff
@@ -0,0 +1,298 @@
+
+/// This file makes heavy use of Copy-on-write references to strings.
+/// Since the logs should be light weight they should not clone any string.
+/// Using generic arguments bound to `Into<Cow<'a, str>>` any reference passed that uses
+/// can provide a reference to a string can be used. E.g. `&String, String, &str, &AccountId`
+/// This also makes it more ergonomic since not only one type can be passed.
+/// 
+/// The unfortunate aspect to this, however, is that the lifetime `'a` is introduced.
+/// This is so that if the `Cow` value has a longer lifetime than the reference it is pointing to it will
+/// copy the string to ensure it still has access to it.  Hence, copy-on-write, where write here is
+/// dropping the reference.
+/// 
+/// While this makes the internal types more complicated the external API will automatically provide the lifetime.
+/// 
+/// See discussion in this PR for more details: https://github.com/near/near-sdk-rs/pull/627
+use std::borrow::Cow;
+
+use serde::{Deserialize, Serialize};
+use serde_with::skip_serializing_none;
+
+#[derive(Serialize, Debug)]
+#[serde(tag = "standard")]
+#[serde(rename_all = "snake_case")]
+pub enum NearEvent<'a> {
+    #[serde(borrow)]
+    Nep171(Nep171Event<'a>),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Nep171Event<'a> {
+    pub version: &'static str,
+    #[serde(flatten)]
+    #[serde(borrow)]
+    pub event_kind: Nep171EventKind<'a>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "event", content = "data")]
+#[serde(rename_all = "snake_case")]
+#[allow(clippy::enum_variant_names)]
+pub enum Nep171EventKind<'a> {
+    #[serde(borrow)]
+    NftMint(Vec<NftMintData<'a>>),
+    #[serde(borrow)]
+    NftTransfer(Vec<NftTransferData<'a>>),
+    #[serde(borrow)]
+    NftBurn(Vec<NftBurnData<'a>>),
+}
+
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NftMintData<'a> {
+    #[serde(borrow)]
+    pub owner_id: Cow<'a, str>,
+    #[serde(borrow)]
+    pub token_ids: Vec<Cow<'a, str>>,
+    #[serde(borrow)]
+    pub memo: Option<Cow<'a, str>>,
+}
+
+impl<'a> NftMintData<'a> {
+    pub fn new<O, T, M>(owner_id: O, token_ids: Vec<T>, memo: Option<M>) -> NftMintData<'a>
+    where
+        O: Into<Cow<'a, str>>,
+        T: Into<Cow<'a, str>>,
+        M: Into<Cow<'a, str>>,
+    {
+        Self {
+            owner_id: owner_id.into(),
+            token_ids: token_ids.into_iter().map(Into::into).collect(),
+            memo: memo.map(Into::into),
+        }
+    }
+}
+
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NftTransferData<'a> {
```

Comment:
Would be nice if there were docs on some of these, but that can always be added in follow up PR